File: ros2_ws/parms/src/cpp_parameters/src/speed_limiter.cpp
#include <chrono>
#include <string>
#include <rclcpp/rclcpp.hpp>

using namespace std::chrono_literals;

class SpeedLimiterNode : public rclcpp::Node
{
public:
  SpeedLimiterNode() : Node("speed_limiter")
  {
    // 1. 建立參數描述
    auto param_desc = rcl_interfaces::msg::ParameterDescriptor{};
    param_desc.description = "機器人的最高移動速限 (m/s)";

    // 2. 宣告參數，名稱為 "max_speed"，預設值為 1.0
    this->declare_parameter("max_speed", 1.0, param_desc);

    // 3. 建立計時器
    timer_ = this->create_wall_timer(
      1000ms, std::bind(&SpeedLimiterNode::timer_callback, this));
  }

private:
  void timer_callback()
  {
    // 4. 取得參數值，並轉型為 double (浮點數)
    double current_speed_limit = this->get_parameter("max_speed").as_double();

    // 印出狀態
    RCLCPP_INFO(this->get_logger(), "目前最高限速設定為: %.2f m/s", current_speed_limit);
  }
  rclcpp::TimerBase::SharedPtr timer_;
};

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<SpeedLimiterNode>());
  rclcpp::shutdown();
  return 0;
}

File: ros2_ws/parms/src/python_parameters/python_parameters/speed_limiter.py
import rclpy
from rclpy.node import Node
from rcl_interfaces.msg import ParameterDescriptor

class SpeedLimiterNode(Node):
    def __init__(self):
        super().__init__('speed_limiter')
        
        # 1. 建立參數描述 (選用，但推薦)
        # 讓其他使用者知道這個參數的用途
        my_desc = ParameterDescriptor(description='機器人的最高移動速限 (m/s)')
        
        # 2. 宣告參數：(參數名稱, 預設值, 描述)
        # 系統會從預設值 '1.0' 自動推斷出這個參數是浮點數 (Double) 型態
        self.declare_parameter('max_speed', 1.0, my_desc)

        # 3. 建立計時器，每 1 秒讀取並印出當前限速
        self.timer = self.create_timer(1.0, self.timer_callback)

    def timer_callback(self):
        # 4. 取得參數的當前值 (加上 .double_value 確保型態正確)
        current_speed_limit = self.get_parameter('max_speed').get_parameter_value().double_value
        
        # 印出狀態
        self.get_logger().info('目前最高限速設定為: %.2f m/s' % current_speed_limit)

def main():
    rclpy.init()
    node = SpeedLimiterNode()
    rclpy.spin(node)
    rclpy.shutdown()

if __name__ == '__main__':
    main()